userCode/awqms/dag.py

```
@@ -0,0 +1,208 @@
+                continue
+
+        resp = requests.post(
+            url_join(API_BACKEND_URL, "Datastreams"),
+            json=datastream.model_dump(by_alias=True),
+        )
+        if not resp.ok:
+            LOGGER.error(f"Failed posting datastream: {datastream}")
+            raise RuntimeError(resp.text)
+
+
+@asset(
+    partitions_def=station_partition, deps=[post_awqms_datastreams], group_name="awqms"
+)
+async def awqms_observations(
+    awqms_metadata: StationData, awqms_datastreams: list[Datastream]
+) -> bool:
+    associatedThing = awqms_metadata.MonitoringLocationId
+
+    observations: dict[int, Observation] = {}
+
+    async def fetch_and_process(datastream: Datastream):
+        observations_ids = fetch_observation_ids(datastream.iotid)
+        LOGGER.info(f"Fetching observations for {datastream.iotid}")
+        for result in fetch_observations(datastream.description, associatedThing):
+            if not result["ResultValue"]:
+                continue
+
+            id = "".join([datastream.iotid, result["StartDateTime"]])
+            iotid = deterministic_hash(id, 18)
+
+            _test = (
+                iotid in observations and result["Status"] != "Final"
+            ) or iotid in observations_ids
+            if _test:
+                continue
+
+            observations[iotid] = to_sensorthings_observation(
+                iotid,
+                datastream,
+                result["ResultValue"],
+                result["StartDateTime"],
+                awqms_metadata.Geometry,
+            )
+
+    # Run fetch_and_process for all datastreams concurrently
+    try:
+        await asyncio.gather(
+            *(fetch_and_process(datastream) for datastream in awqms_datastreams)
+        )
+    except Exception as err:
+        LOGGER.error(err)
+        return False
+
+    awqms_observations = list(observations.values())
+    BatchHelper().send_observations(awqms_observations)
+
+    return True
```

I don't think returning bool is correct here. I think if the observations fail it should raise an error and stop the pipeline.